src/serialization.c

```diff
@@ -279,4 +236,161 @@ RRStatus RaftRedisCommandArrayDeserialize(RaftRedisCommandArray *target, const v
+        if ((n = decodeString(p, buf_size, &target->key_serialized[i])) < 0 || !target->key_serialized[i]) {
+            return RR_ERROR;
+        }
+        p += n; buf_size -= n;
+    }
+
+    return RR_OK;
+}
+
+raft_entry_t *RaftRedisSerializeImport(const ImportKeys *import_keys)
+{
+    int n;
+
+    size_t sz = calcIntSerializedLen(import_keys->term);
+    sz += calcIntSerializedLen(import_keys->migration_session_key);
+    sz += calcIntSerializedLen(import_keys->num_keys);
+    for (size_t i = 0; i < import_keys->num_keys; i++) {
+        sz += calcSerializeStringSize(import_keys->key_names[i]);
+        sz += calcSerializeStringSize(import_keys->key_serialized[i]);
+    }
+
+    /* Prepare entry */
+    raft_entry_t *ety = raft_entry_new(sz);
+    char *p = ety->data;
+
+    /* Encode term */
+    n = encodeInteger('*', p, sz, import_keys->term);
+    RedisModule_Assert(n != -1);
+    p += n; sz -= n;
+
+    /* encode migration_session_key */
+    n = encodeInteger('*', p, sz, import_keys->migration_session_key);
+    RedisModule_Assert(n != -1);
+    p += n; sz -= n;
+
+    /* encode num_keys */
+    n = encodeInteger('*', p, sz, import_keys->num_keys);
+    RedisModule_Assert(n != -1);
+    p += n; sz -= n;
+
+    for (size_t i = 0; i < import_keys->num_keys; i++) {
+        n = encodeString(p, sz, import_keys->key_names[i]);
+        RedisModule_Assert(n != -1);
+        p += n; sz -= n;
+        n = encodeString(p, sz, import_keys->key_serialized[i]);
+        RedisModule_Assert(n != -1);
+        p += n; sz -= n;
+    }
+
+    return ety;
+}
+
+/* serialize out keys in a RaftRedisCommand for locking */
+raft_entry_t *RaftRedisLockKeysSerialize(RedisModuleString **argv, size_t argc)
+{
+    RedisModuleDict * keys = RedisModule_CreateDict(redis_raft.ctx);
+    size_t total_key_size = 0;
+    int num_keys = 0;
+
+    for (size_t idx = 0; idx < argc; idx++) {
+        size_t str_len;
+        RedisModule_StringPtrLen(argv[idx], &str_len);
+
+        if (RedisModule_DictSet(keys, argv[idx], NULL) == REDISMODULE_OK) {
+            total_key_size += str_len + 1;
+            num_keys++;
+        }
+    }
+
+    size_t data_len = calcIntSerializedLen(num_keys) + total_key_size;
+    raft_entry_t *ety = raft_entry_new(data_len);
+    char *p = ety->data;
+
+    /* Encode number of keys */
+    int n = encodeInteger('*', p, total_key_size, num_keys);
+    RedisModule_Assert(n != -1);
+    p += n;
+    data_len -= n;
+
+    RedisModuleDictIter *iter = RedisModule_DictIteratorStartC(keys, "^", NULL, 0);
+    char *key;
+    size_t key_len;
+    while ((key = RedisModule_DictNextC(iter, &key_len, NULL)) != NULL) {
+        RedisModule_Assert(total_key_size >= (key_len + 1));
+        memcpy(p, key, key_len);
+        p += key_len;
+        *p = '\0';
+        p++;
+        data_len -= (key_len + 1);
+    }
+    RedisModule_DictIteratorStop(iter);
+
+    RedisModule_Assert(data_len == 0);
+    RedisModule_FreeDict(redis_raft.ctx, keys);
```

Comment:
```suggestion
    RedisModule_FreeDict(NULL, keys);
```
I asked in some other comment  if this is relevant for us (it is not currently but probably will not be in the future). I'm not sure. Should we pass ctx to these functions as we are not using auto memory?